### PR TITLE
[4.0] Adding version switch in Image and Imagefilter for PHP8

### PR DESCRIPTION
**Pull request by Hackwar**
PHP8 changed the behavior of GD to use objects instead of resources. This introduces version dependent handling in the code and thus should make this work again. Since we have good unittests for this, this should be enough to test this.

### CODE REVIEW

**Comment by richard67** [comment]
@Hackwar Just a question from reading: Ist it correct that the first check if `$source` is true applies only to the first `||` condition in lines 128 and 129?
```
if ($source && (\is_object($source) && get_class($source) == 'GdImage')
	|| (\is_resource($source) && get_resource_type($source) == 'gd'))
```
Or should it apply in both of the cases separated by the `||` as follows:
```
if ($source && (\is_object($source) && get_class($source) == 'GdImage'
	|| \is_resource($source) && get_resource_type($source) == 'gd'))
```
As I said, just a question from reading, maybe all is fine as is it.

**Comment by Hackwar** [comment]
No, that is fine the way it is. $source can be null and get_class(null) (or anything not an object) will throw an error.

Anyhow, with this PR and the bunch of changes we did to drone and the docker images used, I can now happily proclaim that we are passing all tests.

**Comment by zero-24** [comment]
hmm can you think of any reason that RIPS would complain about files in this path: `libraries/vendor/hoa` I don't see that this PR touches anything around that path and is also not introducing new composer deps as well as this path does not seem to be included in with an `composer install`?

**Comment by SniperSister** [comment]
@zero-24 seems to be a RIPS glitch. Whitelisted the issues and restarted the build

**Comment by HLeithner** [comment]
We should be consistent on how we handle this. I think its enough if we check for object and resource (including type) and not for the php version. Also please add a comment that resource can be removed if we only support php 8+

File: libraries/src/Image/Image.php [hunk 1]
```diff
@@ -124,10 +124,14 @@ public function __construct($source = null)
...
-		if (\is_resource($source) && (get_resource_type($source) == 'gd'))
+		/**
+		 * If the source input is a resource, set it as the image handle.
+		 * TODO: Remove check for resource when we only support PHP 8
+		 */
+		if ($source && (\is_object($source) && get_class($source) == 'GdImage')
+			|| (\is_resource($source) && get_resource_type($source) == 'gd'))
 		{
-			$this->handle = &$source;
+			$this->handle = $source;
```
Comment (Hackwar) [inline libraries/src/Image/Image.php:134]: I'm not sure if this by-reference has any effect. In the end we are only handing over a pointer here, right? So it shouldn't matter if we hand this over by reference or not. Can anybody confirm me on that?

**Comment by Hackwar** [comment]
I added the comments and unified the handling. Please review again and give thumbs up/down.

**Comment by HLeithner** [comment]
Thanks

**Pull request merged**